File: integration-tests/TestAPI-bindings/kotlin/src/test/kotlin/com/cricut/testapi/CollectionTests.kt
package com.cricut.testapi

import com.cricut.fishyjoes.runtime.Tuple4
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class CollectionTests {
    @Test
    fun testCollectionValues() {
        assertEquals(Collections.arrayOfInt, arrayListOf(2L, 7L, 3L, 5L, 8L))
        assert(Collections.setOfInt.containsAll(setOf(5L, 9L, 2L, 4L, 3L)))
        assertEquals(Collections.dictionaryOfIntToInt, mapOf(1L to 10L, 2L to 20L, 3L to 30L, 4L to 40L, 5L to 50L))
        assertEquals(Collections.arrayOfBigTuples, arrayListOf<Tuple4<Byte, Short, Int, Long>>(
            Tuple4(1, 2, 3, 4),
            Tuple4(5, 6, 7, 8)
        ));
    }

    @Test
    fun testOptionalCollectionValues() {
        assertEquals(Collections.maybeArrayOfInt, arrayListOf(2L, 7L, 3L, 5L, 8L))
        assertEquals(Collections.maybeSetOfInt?.containsAll(setOf(5L, 9L, 2L, 4L, 3L)), true)
        assertEquals(Collections.maybeDictionaryOfIntToInt, mapOf(1L to 10L, 2L to 20L, 3L to 30L, 4L to 40L, 5L to 50L))
    }

    @Test
    fun testOptionalCollectionOfOptionalValues() {
        assertEquals(Collections.maybeArrayOfMaybeInt, arrayListOf(null, 2L, 7L, 3L, 5L, 8L))
        assertEquals(Collections.maybeSetOfMaybeInt?.containsAll(setOf(null, 5L, 9L, 2L, 4L, 3L)), true)
        assertEquals(Collections.maybeDictionaryOfIntToMaybeInt, mapOf(100L to null, 1L to 10L, 2L to 20L, 3L to 30L, 4L to 40L, 5L to 50L))
    }

    @Test
    fun testFunctionsTakingAndReturningCollectionTypes() {
        assertEquals(Collections.echoArrayOfInt(Collections.arrayOfInt), Collections.arrayOfInt)
        assertEquals(Collections.echoSetOfInt(Collections.setOfInt), Collections.setOfInt)
        assertEquals(Collections.echoDictionaryOfIntToInt(Collections.dictionaryOfIntToInt), Collections.dictionaryOfIntToInt)
    }

    @Test
    fun testFunctionsTakingAndReturningOptionalCollectionTypes() {
        assertEquals(Collections.echoMaybeArrayOfMaybeInt(Collections.maybeArrayOfMaybeInt), Collections.maybeArrayOfMaybeInt)
        assertEquals(Collections.echoMaybeSetOfMaybeInt(Collections.maybeSetOfMaybeInt), Collections.maybeSetOfMaybeInt)
        assertEquals(Collections.echoMaybeDictionaryOfIntToMaybeInt(Collections.maybeDictionaryOfIntToMaybeInt), Collections.maybeDictionaryOfIntToMaybeInt)
    }

    @Test
    fun testFunctionsTakingClosuresWithCollectionTypes() {
        assertEquals(Collections.collectionMapper(arrayListOf(10L, null, 30L)) { (it ?: arrayListOf()).map { (it ?: 1L) * 2L } }, arrayListOf(20L, 2L, 60L))
    }

    @Test
    fun testObjectsWithCollectionMembers() {
        assertEquals(Collections.CollectionHolder.staticProperty, arrayListOf(null, 2L, 7L, 3L, 5L, 8L))
        assertEquals(Collections.CollectionHolder.staticMutableProperty, arrayListOf(null, 2L, 7L, 3L, 5L, 8L))
        Collections.CollectionHolder.staticMutableProperty = arrayListOf(100L, null, 200L)
        assertEquals(Collections.CollectionHolder.staticMutableProperty, arrayListOf(100L, null, 200L))
        val s = Collections.defaultCollectionHolder
        assertEquals(s, Collections.defaultCollectionHolder)
        s.boolDictionary = s.boolDictionary.map { Pair(it.key, !it.value) }.toMap()
        assertNotEquals(s, Collections.defaultCollectionHolder)
    }
}

File: integration-tests/TestAPI/bindings/c-sharp/Cricut.TestAPI.Tests/CollectionTests.cs
using System.Linq;
using Xunit;
using System.Collections.Generic;
using FluentAssertions;

namespace Cricut.TestAPI.Tests {
    [Collection("root")]
    public class CollectionTests {
        [Fact]
        void testCollectionValues() {
            Collections.ArrayOfInt.Should().BeEquivalentTo(new List<nint> { 2, 7, 3, 5, 8 });
            Collections.SetOfInt.Should().BeEquivalentTo(new HashSet<nint> { 5, 9, 2, 4, 3 });
            Collections.DictionaryOfIntToInt.Should().BeEquivalentTo(new Dictionary<nint, nint> { {1, 10}, {2, 20}, {3, 30}, {4, 40}, {5, 50} });
        }

        [Fact]
        void testOptionalCollectionValues() {
            Collections.MaybeArrayOfInt.Should().BeEquivalentTo(new List<nint> { 2, 7, 3, 5, 8 });
            Collections.MaybeSetOfInt.Should().BeEquivalentTo(new HashSet<nint> { 5, 9, 2, 4, 3 });
            Collections.MaybeDictionaryOfIntToInt?.Should().BeEquivalentTo(new Dictionary<nint, nint> { {1, 10}, {2, 20}, {3, 30}, {4, 40}, {5, 50} });
        }

        [Fact]
        void testOptionalCollectionOfOptionalValues() {
            Collections.MaybeArrayOfMaybeInt?.Equals(new List<nint?> {null, 2, 7, 3, 5, 8});
            Collections.MaybeSetOfMaybeInt?.Should().BeEquivalentTo(new HashSet<nint?> {null, 5, 9, 2, 4, 3});
            Collections.MaybeDictionaryOfIntToMaybeInt?.Should().BeEquivalentTo(new Dictionary<nint, nint?> { {100, null}, {1, 10}, {2, 20}, {3, 30}, {4, 40}, {5, 50} });
        }

        [Fact]
        void testFunctionsTakingAndReturningCollectionTypes() {
            Collections.EchoArrayOfInt(Collections.ArrayOfInt).Should().BeEquivalentTo(Collections.ArrayOfInt);
            Collections.EchoSetOfInt(Collections.SetOfInt).Should().BeEquivalentTo(Collections.SetOfInt);
            Collections.EchoDictionaryOfIntToInt(Collections.DictionaryOfIntToInt).Should().BeEquivalentTo(Collections.DictionaryOfIntToInt);
        }

        [Fact]
        void testFunctionsTakingAndReturningOptionalCollectionTypes() {
            Collections.EchoMaybeArrayOfMaybeInt(Collections.MaybeArrayOfMaybeInt)?.Should().BeEquivalentTo(Collections.MaybeArrayOfMaybeInt);
            Collections.EchoMaybeSetOfMaybeInt(Collections.MaybeSetOfMaybeInt).Should().BeEquivalentTo(Collections.MaybeSetOfMaybeInt);
            Collections.EchoMaybeDictionaryOfIntToMaybeInt(Collections.MaybeDictionaryOfIntToMaybeInt).Should().BeEquivalentTo(Collections.MaybeDictionaryOfIntToMaybeInt);
        }

        [Fact]
        void testFunctionsTakingClosuresWithCollectionTypes() {
            Collections.CollectionMapper(new List<nint?> { 10, 20, 30 }, it => new List<nint?>(it!.Select(it => it * 2)))
                .Should().BeEquivalentTo(new List<nint> { 20, 40, 60 });
        }

        [Fact]
        void testObjectsWithCollectionMembers() {
            Collections.CollectionHolder.StaticProperty.Should().BeEquivalentTo(new List<nint?> { null, 2, 7, 3, 5, 8 });
            Collections.CollectionHolder.StaticMutableProperty.Should().BeEquivalentTo(new List<nint?> { null, 2, 7, 3, 5, 8 });
            Collections.CollectionHolder.StaticMutableProperty = new List<nint?> { 100, null, 200 };
            Collections.CollectionHolder.StaticMutableProperty.Should().BeEquivalentTo(new List<nint?> { 100, null, 200 });
            var s = Collections.DefaultCollectionHolder;
            s.Should().BeEquivalentTo(Collections.DefaultCollectionHolder);
            s.BoolDictionary = new Dictionary<bool, bool>(s.BoolDictionary.Select(it => KeyValuePair.Create(it.Key, !it.Value)));
            s.Should().NotBeEquivalentTo(Collections.DefaultCollectionHolder);
        }
    }
}
